### DISCUSSION TITLE
Executing a Python script using rez env and the required version of Python

### DISCUSSION THREAD

**Discussion by sirfuzzalot** [AcademySoftwareFoundation/rez#1113] [Q&A]
**Goal**
Run a Python script (from a rez package) inside a rez environment with the version of Python specified in `requires`.

**Constraints**
- Work on Windows, Linux and MacOS
- executable in one command (ideally)

**What we've tried**

Using an alias to invoke an alias.

```python
# src/tool.py

def main():
    print("Hello World!")


if __name__ == "__main__":
    main()
```

```python
# package.py
name = "cli_tool"
description = "A command line tool"
version = "1.0.0"
requires = ["python-3.8"]


def commands():
    alias("run", "python {root}/src/tool.py")

```
```bash
> rez env cli_tool -- run
```

This results in `python` in ` alias("run", "python {root}/src/cli_tool.py")` being found from PATH rather than the `python-3.8` rez package. In our case Python 2.7 is invoked rather than Python 3.8.

There is a workaround where in the `python-3.8` package we add `env.PATH.prepend("{root}/bin")`. Then `python` will use the Python 3.8 executable. Is there a more idiomatic approach to this goal?

**Comment by hughetop** [comment]
If I understand this correctly, your python-3.8 package.py just needs to add something like this:
```python
def commands():
    import sys
    if sys.platform == 'win32':
        # I haven't gotten aliases to work on windows
        env.PATH.prepend(r'C:\Program Files\Python38')
    elif sys.platform == 'darwin':
        alias('python', '/Library/Frameworks/Python.framework/Versions/3.8/bin/python3')
    else:
        alias('python', '<wherever python-3.8 is installed>')
```
When pyhon-3.8 is included in a rez environment, it's executable is added as an alias (or directory added to the %PATH%). So when you go into a rez environment with python-3.8, `python` will run it's executable.

**Reply by hughetop** [reply]
For clarification, I edited that package.py by hand so it would handle all operating systems. For me, the `rez-bind python` didn't work because of permissions on windows so I don't know what the typical python package.py file looks like.

**Reply by instinct-vfx** [reply]
Your python packages should account for adding to PATH. That's what it is supposed to do. However i would strongly suggest to not make the platform distinction in the package and instead create variants for the different platform and have them specific to the platform.

**Reply by hughetop** [reply]
I'm a bit confused but I created a separate discussion for it: https://github.com/AcademySoftwareFoundation/rez/discussions/1330.

**Comment by instinct-vfx** [comment]
Could you share your python package? It seems that might be the culprit. If you add your python rez package paths to PATH in your package that should win over 2.7 on PATH as all rez package appended PATH entries are prepended to the system PATH.
